Improve Session Management docs (#1735)
viamrobotics/docs: docs/program/apis/sessions.md

---
title: "Session Management"
linkTitle: "Session Management"
weight: 20
type: "docs"
description: "Manage sessions between your robot and clients connected through Viam's SDKs."
tags: ["client", "sdk", "viam-server", "networking", "apis", "robot api", "session", "sessions", "session management"]
---

When you connect to a robot using an SDK, the SDK connects to the robot's `viam-server` instance as a _client_.
The period of time during which a client is connected to a robot is called a *session*.

*Session management* is a presence mechanism that allows you to manage the clients that are authenticated and communicating with a robot's `viam-server` instance.

> As a safety precaution, the default Session management configuration ensures that a robot only moves when a client is actively connected.

Reviewer: ```suggestion
As a safety precaution, the default session management configuration ensures that a robot only moves when a client is actively connected.
```

Reviewer: Lowercase sessions

This is especially important for robots that physically move.
For example, imagine a wheeled rover gets a [`SetPower()`](/components/base/#setpower) command as the last input from a client before the connection to the robot is interrupted.
Without session management, the API request from the client would cause the rover's motors to move, causing the robot to continue driving forever and potentially colliding with objects and people.

> If you need to different functionality, you can manage sessions yourself.

Reviewer: ```suggestion
If you want to manage operations differently, you can manage sessions yourself.
```

Reviewer: (opt) Phrasing

The Session Management API provides functionality for

- clients to notify to the robot that the client is actively authenticated and connected
- the robot to stop moving components when a session ends

### The `SessionsClient`

> A *client* of a Viam robot can be an SDK script controlling the robot, an input controller, or the different resources on the robot, including {{< glossary_tooltip term_id="part" text="parts" >}} and sub-parts, communicating.

Reviewer: ```suggestion
A *client* of a Viam robot can be an SDK controlling the robot, an input controller, or the different resources on the robot, including {{< glossary_tooltip term_id="part" text="parts" >}} and sub-parts, communicating.
```
(opt) I think either one is somewhat accurate but I don't want to confuse people otherwise by referring to SDK as client / vs just scripts

Author: ```suggestion
A *client* of a Viam robot can be a program using an SDK to control the robot, an input controller, or the different resources on the robot, including {{< glossary_tooltip term_id="part" text="parts" >}} and sub-parts, communicating.
```

Viam's session management API's `SessionsClient` is a built-in solution that manages the connection between your robot's clients and your robot.
If you connect to your robot using one of Viam's SDKs, the resulting client will automatically maintain the session by sending a *heartbeat* notifying the robot's `viam-server` instance of its continued presence.
The `SessionsClient` on the robot maintains an overview of all sessions based on the clients' heartbeat messages.

If the robot does not receive a signal from the client in the expected interval, the robot ends the session and stop all resources that are marked for safety monitoring and have been last used by that session.
That means if a client sends a command to a robot to move the motors and then loses the connection, the robot will stop moving.

{{< alert title="Caution" color="caution" >}}
If a resource was used by client A and then by client B, and client A disconnects, the resource will not be stopped, regardless of possibly ongoing commands initiated by client A.
{{< /alert >}}

> A disconnected client will attempt to establish a new session immediately prior to the next operation it performs.

Reviewer: This part looks great! Thank you for updating this naomi!

linkTitle: "Session Management"
weight: 20
type: "docs"
description: "Manage sessions between your robot and clients connected through Viam's SDKs."
tags: ["client", "sdk", "viam-server", "networking", "apis", "robot api", "session", "sessions", "session management"]
---

When you connect to a robot using an SDK, the SDK connects to the robot's `viam-server` instance as a _client_.
The period of time during which a client is connected to a robot is called a _session_.

_Session management_ is a presence mechanism that allows you to manage the clients that are authenticated and communicating with a robot's `viam-server` instance.

> As a safety precaution, the default session management configuration ensures that a robot only moves when a client is actively connected.

Reviewer: Really like this phrasing!

Author: thank you!

This is especially important for robots that physically move.
For example, imagine a wheeled rover gets a [`SetPower()`](/components/base/#setpower) command as the last input from a client before the connection to the robot is interrupted.
Without session management, the API request from the client would cause the rover's motors to move, causing the robot to continue driving forever and potentially colliding with objects and people.

> If you want to manage operations differently, you can manage sessions yourself.

Reviewer: ```suggestion
If you want to manage operations differently, you can manage your robot's client sessions yourself.
```
(opt) to make this extra clear

The Session Management API provides functionality for

- clients to notify to the robot that the client is actively authenticated and connected
- the robot to stop moving components when a session ends

### The `SessionsClient`

> A _client_ of a Viam robot can be a program using an SDK to control the robot, an input controller, or the different resources on the robot, including {{< glossary_tooltip term_id="part" text="parts" >}} and sub-parts, communicating.

Reviewer: ```suggestion
A _client_ of a Viam robot can be a program using an SDK to control the robot, or all the different resources on the robot,  including all {{< glossary_tooltip term_id="part" text="parts" >}} and sub-parts, like an input controller and a base, communicating.
```
(opt) Not sure what could be best but I think this sentence could be improved bc as is it kind of seems to separate input controller, resources, and parts, when they are all different names that could apply to an input controller hardware or base hardware etc. within the viam platform

> For example, if you use Viam's module registry to [add modular resources to your robot](/extend/modular-resources/), the clients of your robot will include the model servers you instantiate on your robot for individual resources.

Reviewer: ```suggestion
For example, if you use Viam's module registry to [add modular resources to your robot](/extend/modular-resources/), the clients of your robot will include the model servers you instantiate on your robot for individual resources, as well as whatever SDK(s) you are using to program the modular resources.
```
(opt)

> For example, if you use Viam's module registry to [add modular resources to your robot](/extend/modular-resources/), the clients of your robot will include the model servers you instantiate on your robot for individual resources, as well as the SDK(s) you are using to program the modular resources.

Author: ```suggestion
For example, if you use Viam's module registry to [add modular resources to your robot](/extend/modular-resources/), the clients of your robot will include the model servers you instantiate on your robot for individual resources, as well as the SDKs you are using to program the modular resources.
```

The default heartbeat window is 2 seconds if this configuration is omitted.

If you manually manage sessions, each client must send at least one heartbeat within the window you set.

## Manage sessions with the session management API

> The session management API is not currently provided in the Python or TypeScript SDKs.

Author: ```suggestion
The [Session Management API](https://pkg.go.dev/go.viam.com/rdk/session) is not currently provided in the Python or TypeScript SDKs.
```

Use the Go Client SDK instead.

> To manage your sessions with the session management API:

Reviewer: ```suggestion
To manage your session with the session management API:
```
(opt) Maybe change to singular, re. past feedback just to keep all the same (it's hard to just say session or sessions though)

To manage sessions on-robot manually, you can use the following client SDKs:

{{< tabs >}}
{{% tab name="Go" %}}

To enable the [Session Management API](https://pkg.go.dev/go.viam.com/rdk/session) the Go Client SDK provides, [disable the default behavior of sessions](#disable-default-session-management).
If you manually manage sessions, each client must send at least one heartbeat within the window you set.

{{% /tab %}}
{{% tab name="Other SDKs" %}}
## Manage sessions with the session management API

The session management API is not currently provided in the Python or TypeScript SDKs.
The [Session Management API](https://pkg.go.dev/go.viam.com/rdk/session) is not currently provided in the Python or TypeScript SDKs.
Use the Go Client SDK instead.

{{% /tab %}}
{{% /tabs %}}

### Use the session management API to manually manage sessions
To manage your sessions with the session management API:

First, use your [`RobotClient()`](/program/apis/#robot-api) instance to access the [`SessionsClient`](https://pkg.go.dev/go.viam.com/rdk/session) within your client SDK program.
This is a [gRPC](https://grpc.io/) client that `viam-server` instantiates at robot runtime.

> Find `SessionsClient` defined on [GitHub](https://github.com/viamrobotics/rdk/blob/main/robot/client/client.go).

Reviewer: Fine with taking this out but AFAIK the rdk docs do not include everything in GitHub-- some methods are not documented or are missing comments in code (like expireLoop in SessionManager)

Author: This has just moved lower on the page! All still there :)

This option allows you to have full control over sessions management.
After disabling the client, you must now manage each of your sessions manually with the session management API.
You can do this with Viam's [client SDKs](https://pkg.go.dev/go.viam.com/rdk/session).

> 

Author: ```suggestion
```
